### PR TITLE
Added PendingFutureLimiter and ProfiledPendingFutureLimiter

### PR DESCRIPTION
**Pull request by surkovich**

### CODE REVIEW

File: jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java
Comment (swarmshine) [inline jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java:14]: Since we are publishing this code to github it will be better to use English for our documentation

File: jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java
Comment (swarmshine) [inline jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java]: maxFutureExecuteTimeout can be 0, it is better to mention that constructor documentation

File: jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java
Comment (swarmshine) [inline jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java]: Not sure about `Trespassing`, trespassing infer violation, but here we do not have any offence against our rules.  
Suggestion: leave name setThresholdListener() and in listener name callbacks methods as 
 onHiLimitReached  and onLowLimitSubceed

File: jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java
Comment (swarmshine) [inline jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java:201]: There is a not  straight forward behavior in   awaitAndPurge, releaseAll, releaseEnqueued in terms of maxFutureExecuteTime.  
releaseAll, releaseEnqueued know about maxFutureExecuteTime but don't know that it can be 0. 
awaitAndPurge knows about maxFutureExecuteTime and that it can be 0 and also knows that releaseEnqueued do not check if maxFutureExecuteTime is 0.  

suggestion:

releaseTimeoutedFutures() - if maxFutureExecuteTime is 0 do not release anything and immediately returns, otherwise release timeouted futures. 

releaseAll rename to waitAll() - waits for all futures to complete normally or by timeout

awaitAndPurge rename to awaitOpportunityToEnqueueAndPurge

**Review by swarmshine — CHANGES_REQUESTED** [review]

File: jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java
Comment (surkovich) [reply jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java:201]: > releaseAll rename to waitAll() - waits for all futures to complete normally or by timeout

Actual behavior of  this this method is not to wait, but just to release all queued futures. If we'll change the releaseEnqueued to releaseTimeoutedFutures and will exit if maxFutureExecuteTime is 0 it still will not wait for anything. 
I could change it to wait for all futures to complete, but i'm not sure if it makes sense, and how we'll do that? Just block the current thread until all the futures will be completed? 
Suggestion about this method - rename it to the releaseAllTimeoutedIfPossible().

About the rest comment - totally agree.

Comment (surkovich) [reply jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java:201]: Final decision - we need a method(s), which will wait all futures to complete successfully or to wait till timeout and then release. Thus, discussed method (reaseAll()) will be replaced by 
waitAll();//block invoked thread and wait all futures to complete
waitAll(long timeoutMillis); //block invoked thread and wait all futures to complete or till timeout is reached.

File: jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java
Comment (swarmshine) [inline jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java]: better to translate documentation to english

**Review by swarmshine — CHANGES_REQUESTED** [review]
other change request discussed verbally

File: jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java
Comment (swarmshine) [inline jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java]: Maybe releaseTimeoutedIfPossible(); should be invoked within loop ?

**Review by swarmshine — CHANGES_REQUESTED** [review]

File: jfix-stdlib-concurrency/src/test/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiterTest.java
Comment (swarmshine) [inline jfix-stdlib-concurrency/src/test/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiterTest.java]: Your test shows you that it is a good idea to allow user to configure waitTimeToCheckSizeQueue.   
But you are ignoring it and hacking access to the field via reflection.    
Suggestion: 
1. rename waitTimeToCheckSizeQueue to pendingQueueSizeChangeCheckInteval
2. add this parameter to constructor or via public setter. 
This way no reflection should be required

File: jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java
Comment (swarmshine) [inline jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java]: counter.wait(Math.min(pendingQueueSizeChangeCheckInteval, timeoutMillis - System.currentTime()-startTime))

File: jfix-stdlib-concurrency/src/test/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiterTest.java
Comment (swarmshine) [inline jfix-stdlib-concurrency/src/test/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiterTest.java]: Why you are  creating new thread that consume resources just to create CompletableFuture that you can control and complete on demand.  You can simply 
```
val latch = CompletableFuture()
createTask(){
    return latch.thenApply{}
}
unleashLatchAndCompleteAllTask(){ latch.complete() }
...

File: jfix-stdlib-concurrency/src/test/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiterTest.java
Comment (swarmshine) [inline jfix-stdlib-concurrency/src/test/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiterTest.java:44]: it is unclear to the reader that maxPendingCount is 3. Suggestion: in test make explicit configuration 
```
PendingFutureLimiter limiter = new LimiterBuilder()
                .maxPendingCount(3)
                .enqueueTasks(3)
```

**Review by swarmshine — CHANGES_REQUESTED** [review]

File: jfix-stdlib-concurrency/src/test/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiterTest.java
Comment (surkovich) [reply jfix-stdlib-concurrency/src/test/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiterTest.java]: Constructor already has two parameters, well, one more and we'll reach the sane limit... Therefore it seems less evil to me to add a public setter, it is volatile anyway and doesn't seem to crush any functionality.

File: jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java
Comment (surkovich) [reply jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java]: Agree.

File: jfix-stdlib-concurrency/src/test/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiterTest.java
Comment (surkovich) [reply jfix-stdlib-concurrency/src/test/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiterTest.java]: Nice trick... Applied.

File: jfix-stdlib-concurrency/src/test/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiterTest.java
Comment (surkovich) [reply jfix-stdlib-concurrency/src/test/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiterTest.java:44]: Added

File: jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java
Comment (surkovich) [reply jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java]: Translated

File: jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java
Comment (surkovich) [reply jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java]: Mentioned

**Pull request merged**
